ENH: Adds range types

llllllllll opened the pull request:
I wanted to add postgres RANGE types to blaze so I needed some type to represent the structure.

cpcloud commented on datashape/discovery.py:
```diff
@@ -396,3 +396,29 @@ def descendents(d, x):
+
+
+def _tp_eq(a, b):
+    return a == b or b is null
```
not sure i understand the logic here. if `b` is `null` then `a` is equal to `b`?

llllllllll replied:
This is probably a bad name. The idea is that `null` represents an an infinite bound, or default step, so this was used to check that all the types for start, stop, and step are the same or null.

cpcloud replied:
i'd probably just inline this code with a short comment explaining what it means

cpcloud replied:
If you raise on null, then you can just inline the `a == b` in the lambda below

llllllllll replied:
I don't think we want to raise on NULL, the idea is that slice(None, 5) should be a range of `int` even though one side is unbound.

cpcloud commented on datashape/tests/test_coretypes.py:
```diff
@@ -315,3 +316,15 @@ def test_duplicate_field_names_fails():
+     ('string[10, "A"]', String(10, 'A'))],
+)
+def test_range(ds):
+    assert dshape('range[{d}]'.format(d=ds[0])) == dshape(str(Range(ds[1])))
```
what does `Range(string[10, "A"])` mean?

cpcloud replied:
probably should test against `date` and `datetime` here

llllllllll replied:
we can take the range over any sortable thing, so for example `slice('a', 'z')`

I can add a test for dates

cpcloud replied:
i understand the use case for slicing single characters, but having slicing defined for all strings of length 10 doesn't seem that useful. Am I missing something here?

llllllllll replied:
I guess I was trying to test a subtype with a parameter. Could someone not provide a range of say, column names that would be expressed as range[string]?

cpcloud replied:
possibly, but i think that merits a bit more discussion. let's hold off on that for now by raising a NotImplementedError and discuss more after merging this PR

cpcloud commented on datashape/tests/test_discovery.py:
```diff
@@ -292,3 +294,15 @@ def __init__(self, shape, dtype):
+
+def test_discover_slice():
+    assert discover(slice(1)) == Range(int64)
+    assert discover(slice(None)) == Range(float64)  # default
```
any reason for this particular default? i'm a little -1 on having a default at all, as i think if someone is going to use the `Range` type then they need to specify a type

llllllllll replied:
I guess it is common to use an unbound range (-infinity,infinity) so I wasn't sure what the type of that would be. Maybe just don't discover slices with no types?

cpcloud replied:
yeah, I'd raise an error if a slice has no types or mixed types

cpcloud commented:
@llllllllll looks like merge conflicts. likely from my dead code cleanup pr

cpcloud commented:
@llllllllll what is the status on this?

cpcloud commented on datashape/discovery.py:
```diff
@@ -405,3 +405,29 @@ def descendents(d, x):
+            break
+    else:
+        # empty slice discovers to 'float64'
+        return Range(float64)
```
I think you should raise here instead of defaulting to `float64`

llllllllll commented:
updated and rebased. The status of this was that I wrote it because I was working with range types but I decided to change the underlying implementation anyway. There is still a decent chunk of work that needs to be done in blaze and odo to work with this so I was sort of just putting it off. This can of course go into datashape before we are ready to add the support to blaze and odo because it will make it easier for us to do that later.
